### PR TITLE
Add component argument inference

### LINKED ISSUE #751: Infer the arguments based on component `__init__` arguments
When the user uses Lightweight Python components (#558) we want to get any information we currently get from the component spec from the provided Python code.

For the arguments, we should be able to get them from the `__init__` arguments of the component class. These should be typed to provide the proper information, but we can fall back on string type for any untyped arguments.

### PR DESCRIPTION
**Pull request by RobbeSneyders**
Fixes #751 

This PR introduces functionality to infer the arguments from a `Component` class. The result is a dictionary with the argument names as keys, and `Argument` instances as values, which is the format of [`component_spec.args`.](https://github.com/ml6team/fondant/blob/8e828441eec8ff91074e5c8ccf16fe405b719594/src/fondant/core/component_spec.py#L193)

We can leverage this behavior for Lightweight Python components as described in #750.

Did some TDD here, let me know if I missed any cases.

### CODE REVIEW

**Comment by RobbeSneyders** [comment]
Pipeline failure is unrelated. Submitted #764 to fix it.

File: tests/pipeline/test_argument_inference.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        "list_argument": Argument(
+            name="list_argument",
+            type=list,
+            optional=True,
+            default=None,
+        ),
+    }
+
+
+def test_parametrized_types_old():
```
Comment (PhilippeMoussalli) [inline tests/pipeline/test_argument_inference.py:144]: what's the "old" referring to? can we keep it to `test_parametrized_types` to match the above?

**Review by PhilippeMoussalli — APPROVED** [review]
Thanks Robbe! Nice work, the tests are quite extensive and seem to cover all scenarios

File: src/fondant/pipeline/argument_inference.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import typing as t
+
+from fondant.component import Component
+from fondant.core.component_spec import Argument
+from fondant.core.exceptions import UnsupportedTypeAnnotation
+
+BUILTIN_TYPES = [str, int, float, bool, dict, list]
+
+
+def annotation_to_type(annotation: t.Any) -> t.Type:
```
Comment (GeorgesLorre) [inline src/fondant/pipeline/argument_inference.py:11]: Nice !

File: tests/pipeline/test_argument_inference.py
Comment (GeorgesLorre) [inline tests/pipeline/test_argument_inference.py:1]: 😍

**Review by GeorgesLorre — APPROVED** [review]

**Pull request merged**